### PR TITLE
refactor(core): blanket implementation of connection upgrade

### LINKED ISSUE #4307: core: introduce `{In,Out}boundConnectionUpgrade` traits
The `InboundUpgrade` and `OutboundUpgrade` traits currently fulfill two needs:

- Upgrading connections: i.e. going from an unencrypted connection to an encrypted connection with `libp2p-noise` or `libp2p-tls` but also going from a single-stream connection to a multiplexed connection using `libp2p-yamux` or `libp2p-mdns`
- Upgrading streams: i.e. going from a negotiated stream to some kind of message. This isn't always possible which is why we want to remove the concept of stream upgrades to allow for a more uniform handling. See #2863 for details.

To finish #2863, we need to deprecate the implementations of `{In,Out}boundUpgrade` for stream upgrades but without affecting implementations of connection upgrades.

Thus, we need to create a dedicated set of traits that is used just for connection upgrades: `{In,Out}boundConnectionUpgrade`. They need a blanket impl over everything that implements `{In,Out}boundUpgrade` for this to be backwards-compatible.

We can then change the "transport upgrade" infrastructure to require these traits instead of the current ones. For example here: https://github.com/libp2p/rust-libp2p/blob/02dc432227ee1132b7fe8f0b849fc9d513c907d5/core/src/transport/upgrade.rs#L104-L105

For it actually be a backwards-compatible change, we have to rely on the blanket impl in all the dependent crates, like `libp2p-noise` etc. Changing those to the new traits will have to be in the next breaking release.

### PR DESCRIPTION
**Pull request by PanGan21**
## Description

Introduces blanket implementation of `{In,Out}boundConnectionUpgrade` and uses it in transport upgrade infrastructure.

Resolves: #4307.

## Change checklist

- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] A changelog entry has been made in the appropriate crates

### CODE REVIEW

**Comment by mxinden** [comment]
Just approved CI to run.

File: core/src/upgrade.rs
Comment (mxinden) [inline core/src/upgrade.rs]:
```suggestion
    fn upgrade_outbound(self, socket: T, info: Self::Info) -> Self::Future;
```

File: core/src/upgrade.rs
Comment (mxinden) [inline core/src/upgrade.rs]:
```suggestion
    fn upgrade_inbound(self, socket: T, info: Self::Info) -> Self::Future;
```

**Review by mxinden — COMMENTED** [review]
Good work. Thank you.

Next steps for this pull request would be:

- Deprecate the existing `{Out,In}boundUpgrade` trait definitions.
- Bump `libp2p-core` patch version.
- Opposite to what @thomaseizinger suggests in https://github.com/libp2p/rust-libp2p/issues/4307 I think we can already update the `transports/` and `muxers/` protocols to the new `{Out,In}boundConnectionUpgrade`s. Given the blanket implementation, this is a backwards compatible change. Am I missing something @thomaseizinger?

**Comment by thomaseizinger** [comment]
> * Given the blanket implementation, this is a backwards compatible change. Am I missing something @thomaseizinger?

It is not. Removing a trait implementation for a public struct is a breaking change. Imagine someone having a function:

```rust
fn bar(foo: impl InboundUpgrade);
```

Currently, calling this function with `noise::Config` will compile. If we remove the implementation for `InboundUprade`, it will stop compiling, hence it is a breaking change.

**Comment by thomaseizinger** [comment]
> Deprecate the existing `{Out,In}boundUpgrade` trait definitions.

Should we already do that in here? I will require us spread a lot of `allow(deprecated)` over the code. I'd like to first have things like `futures-bounded` in place to actually allow users to migrate their protocol implementations. We will also have to write a detailed migration guide for this I think.

In order to not block this PR on it, I'd suggest we don't deprecate this in here.

File: core/src/upgrade.rs
Comment (PanGan21) [reply core/src/upgrade.rs]: Oops. fixed

File: core/src/upgrade.rs
Comment (PanGan21) [reply core/src/upgrade.rs]: fixed as well

**Comment by PanGan21** [comment]
> 

Let me know what is required. I would be open to continue this work in a separate pr (with some guidance ofc)

**Comment by PanGan21** [comment]
> > Deprecate the existing `{Out,In}boundUpgrade` trait definitions.
> 
> Should we already do that in here? I will require us spread a lot of `allow(deprecated)` over the code. I'd like to first have things like `futures-bounded` in place to actually allow users to migrate their protocol implementations. We will also have to write a detailed migration guide for this I think.
> 
> In order to not block this PR on it, I'd suggest we don't deprecate this in here.


Let me know what is required. I would be open to continue this work in a separate pr (with some guidance ofc)

**Comment by mxinden** [comment]
> > * Given the blanket implementation, this is a backwards compatible change. Am I missing something @thomaseizinger?
> 
> It is not. Removing a trait implementation for a public struct is a breaking change. Imagine someone having a function:
> 
> ```rust
> fn bar(foo: impl InboundUpgrade);
> ```
> 
> Currently, calling this function with `noise::Config` will compile. If we remove the implementation for `InboundUprade`, it will stop compiling, hence it is a breaking change.

Oh, yes, quite obviously. You are right.

**Comment by mxinden** [comment]
> > * Given the blanket implementation, this is a backwards compatible change. Am I missing something @thomaseizinger?
> 
> It is not. Removing a trait implementation for a public struct is a breaking change. Imagine someone having a function:
> 
> ```rust
> fn bar(foo: impl InboundUpgrade);
> ```
> 
> Currently, calling this function with `noise::Config` will compile. If we remove the implementation for `InboundUprade`, it will stop compiling, hence it is a breaking change.

Oh, yes, quite obviously. You are right. Thank you.

**Comment by mxinden** [comment]
> > Deprecate the existing `{Out,In}boundUpgrade` trait definitions.
> 
> Should we already do that in here? I will require us spread a lot of `allow(deprecated)` over the code. I'd like to first have things like `futures-bounded` in place to actually allow users to migrate their protocol implementations. We will also have to write a detailed migration guide for this I think.
> 
> In order to not block this PR on it, I'd suggest we don't deprecate this in here.

:+1: makes sense. Thank you for expanding @thomaseizinger.

**Comment by mxinden** [comment]
> Let me know what is required. I would be open to continue this work in a separate pr (with some guidance ofc)

Thank you @PanGan21. Let's get this pull request merged first. There are still multiple `{In,Out}boundUpgrade` usages in `core/`, e.g. `core/src/upgrade/select`. The only usage of `{In,Out}boundUpgrade` in `core/` should be in the blanket implementation for `{In,Out}boundConnectionUpgrade`.

**Comment by PanGan21** [comment]
> > Let me know what is required. I would be open to continue this work in a separate pr (with some guidance ofc)
> 
> Thank you @PanGan21. Let's get this pull request merged first. There are still multiple `{In,Out}boundUpgrade` usages in `core/`, e.g. `core/src/upgrade/select`. The only usage of `{In,Out}boundUpgrade` in `core/` should be in the blanket implementation for `{In,Out}boundConnectionUpgrade`.

By doing this and replacing the usage of the previous traits with the new ones I am getting error for conflicting implementations. Am I missing something important?
E.g.:
```
error[E0119]: conflicting implementations of trait `OutboundConnectionUpgrade<_>` for type `DeniedUpgrade`
   --> core/src/upgrade.rs:176:1
    |
176 | impl<U, T> OutboundConnectionUpgrade<T> for U
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ conflicting implementation for `DeniedUpgrade`
    |
   ::: core/src/upgrade/denied.rs:50:1
    |
50  | impl<C> OutboundConnectionUpgrade<C> for DeniedUpgrade {
    | ------------------------------------------------------ first implementation here
    |
    = note: downstream crates may implement trait `upgrade::OutboundUpgrade<_>` for type `upgrade::denied::DeniedUpgrade`
```

**Comment by thomaseizinger** [comment]
> > Let me know what is required. I would be open to continue this work in a separate pr (with some guidance ofc)
> 
> Thank you @PanGan21. Let's get this pull request merged first. There are still multiple `{In,Out}boundUpgrade` usages in `core/`, e.g. `core/src/upgrade/select`. The only usage of `{In,Out}boundUpgrade` in `core/` should be in the blanket implementation for `{In,Out}boundConnectionUpgrade`.

I am not sure that is correct. Most implementations of `{In,Out}boundUpgrade` in `libp2p-core` are not actually for connection upgrades but for stream upgrades, e.g. `DeniedUpgrade` etc. I think the only one that is useful for connections is `SelectUpgrade` (to compose muxers or encryption protocols).

I think the way we have to handle this for now is to tag all those with `#[allow(deprecated)]`. Once we remove `{In,Out}boundUpgrade`, most of those upgrades will simply be deleted because they no longer serve a purpose. I envision that once we land https://github.com/libp2p/rust-libp2p/pull/4120, users should no longer need to compose any connection upgrades themselves anyway, meaning we can probably delete most of them or make them implementation details of the new `SwarmBuilder`.

**Comment by mxinden** [comment]
> I think the way we have to handle this for now is to tag all those with `#[allow(deprecated)]`.

Sounds good to me. @PanGan21 sorry for the confusion.

> I envision that once we land #4120, users should no longer need to compose any connection upgrades themselves anyway, meaning we can probably delete most of them or make them implementation details of the new `SwarmBuilder`.

I don't think the first iteration of #4120 can cover all cases and thus I don't think we can remove the `libp2p_core::upgrade` infrastructure any time soon. Though I don't think that decision is relevant for this pull request.

**Comment by thomaseizinger** [comment]
> > I envision that once we land #4120, users should no longer need to compose any connection upgrades themselves anyway, meaning we can probably delete most of them or make them implementation details of the new `SwarmBuilder`.
> 
> I don't think the first iteration of #4120 can cover all cases and thus I don't think we can remove the `libp2p_core::upgrade` infrastructure any time soon. Though I don't think that decision is relevant for this pull request.

Maybe not in the first iteration but hopefully at some point :)

If we keep it, we could at least move it to some kind of util package that isn't part of the public API of `libp2p` such that we can more easily make changes to it without affecting the majority of users.

**Comment by PanGan21** [comment]
Hi guys. I am just wondering, is there something I can do more to have this pr done? Or it is targeting a later release so I have to wait?

**Review by thomaseizinger — APPROVED** [review]
Sorry that this has been dragging on for so long! Normally we don't have such a long turn-around time for PRs but this one somehow slipped through!

It looks good to me! I'll write up a follow-up issue on what needs to be done next! :)

**Pull request merged**
